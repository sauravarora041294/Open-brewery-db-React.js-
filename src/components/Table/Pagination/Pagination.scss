.PaginationSection {
    margin-top: 8px;
    
    .PaginateButton {
        padding: 8px;
        margin: 8px;
        // width: px;
    }
}