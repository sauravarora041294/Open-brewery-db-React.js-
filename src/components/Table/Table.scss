.TableHeader, .TableBody {
    tr th {
        text-align: left;
    }

    tr{
        display: flex;
        margin-bottom: 8px;
        
        & > :first-child{
            flex: 0 0 40%;
        }
        & > :nth-child(2) {
            flex: 0 0 20%;
        }

        & > :nth-child(3) {
            flex: 0 0 20%;
        }

        & > :nth-child(4) {
            flex: 0 0 20%;
        }

        & > :nth-child(5) {
            flex: 0 0 40px;
        }
    }
}