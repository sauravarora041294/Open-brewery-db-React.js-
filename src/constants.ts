export const DASHBOARD_ITEMS_PER_PAGE = 10;