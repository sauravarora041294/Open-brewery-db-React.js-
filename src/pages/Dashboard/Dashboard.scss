.MainSection {
  padding: 8px;
  
  .TableSection {
    display: flex;
    flex-direction: row;
    
    .TableWrapper {
      flex: 0 0 60%;
    }

    .FormWrapper {
      flex: 0 0 40%;
    }
  }
}