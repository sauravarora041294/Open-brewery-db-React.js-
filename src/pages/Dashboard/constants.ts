export const DashboardTableColumns = [
    "Brewery Name",
    "Brewery Type",
    "City",
    "Country",
    "Rating",
    ""
]