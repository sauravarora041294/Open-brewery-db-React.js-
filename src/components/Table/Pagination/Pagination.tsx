import { FC, ReactElement } from "react";
import './Pagination.scss';

interface Props {
    currentPage: number;
    handlePageChange: (counter: number) => void;
}

const Pagination: FC<Props> = ({currentPage, handlePageChange}): ReactElement => {
    return <div className="PaginationSection">
        <button className="PaginateButton" disabled={currentPage === 1} onClick={() => handlePageChange(-1)}>Previous</button>
        <span>{currentPage}</span>
        <button className="PaginateButton" onClick={() => handlePageChange(1)}>Next</button>
    </div>
}

export default Pagination;