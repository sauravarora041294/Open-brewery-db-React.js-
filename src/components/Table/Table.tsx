import { FC, ReactElement, ReactNode } from "react";
import { IDashboard } from "../../interfaces/dashboard";
import { DashboardTableColumns } from "../../pages/Dashboard/constants";
import "./Table.scss";

interface Props {
    tableData: IDashboard[];
    onEdit: (id: string) => void;
    columns: string[];
    children: JSX.Element;
}

const Table: FC<Props> = ({tableData, onEdit, columns, children}): ReactElement => {
    return (
        <table className="Table">
            <thead className="TableHeader">
            <tr>
                {columns.map((columnItem: string) => <th>{columnItem}</th>)}
            </tr>
            </thead>
            <tbody className="TableBody">
                {children}
            </tbody>
        </table>
    )
}

export default Table;