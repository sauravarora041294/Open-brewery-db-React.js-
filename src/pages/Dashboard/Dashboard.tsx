import { ReactElement } from 'react';
import Pagination from '../../components/Table/Pagination/Pagination';
import Table from '../../components/Table/Table';
import Form from './Form/Form';
import './Dashboard.scss';
import { useDashboardData, usePagination } from './hooks';
import { DashboardTableColumns } from './constants';
import { IDashboard } from '../../interfaces/dashboard';

const Dashboard = (): ReactElement => {
  const {currentPage, handlePageChange} = usePagination();
  const {paginatedData, onEdit, onAdd, editData, isEdit, handleEdit} = useDashboardData(currentPage);

  return (
    <div className="MainSection">
      <h2>Open Brewery DataBase</h2>
      <div className="TableSection">
          <div className="TableWrapper">
            <Table tableData={paginatedData} columns={DashboardTableColumns} onEdit={(id: string) => onEdit(id)}>
                  <>
                  {paginatedData.map((tableItem: IDashboard) => {
                    return (<tr key={`tableRow-${tableItem.id}`} >
                    <td>{tableItem.name}</td>
                    <td>{tableItem.brewery_type}</td>
                    <td>{tableItem.city}</td>
                    <td>{tableItem.country}</td>
                    <td>{tableItem.rating ?? "0"}</td>
                    <td>
                        <button onClick={() => onEdit(tableItem.id)}>Edit</button>
                    </td>
                </tr>)
                })}
                </>
            </Table>
            <Pagination currentPage={currentPage} handlePageChange={handlePageChange} />
          </div>
          <div className="FormWrapper">
            <Form data={editData} onAdd={onAdd} isEdit={isEdit} handleEdit={handleEdit}/>
          </div>
      </div>
    </div>
  );
}

export default Dashboard;
