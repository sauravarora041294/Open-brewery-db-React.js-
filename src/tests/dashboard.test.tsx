import { render, screen } from '@testing-library/react';
import Dashboard from '../pages/Dashboard/Dashboard';
import Table from '../components/Table/Table';
import {shallow} from "enzyme";
import Pagination from '../components/Table/Pagination/Pagination';
import Form from '../pages/Dashboard/Form/Form';
import { mockPagedData } from './data';
import { DashboardTableColumns } from '../pages/Dashboard/constants';

describe("Dashboard page", () => {
  test('renders page heading', () => {
    render(<Dashboard />);
    const header = screen.getByText(/Open Brewery DataBase/i);
    expect(header).toBeInTheDocument();
  });

  test("it should render dashboard component", () => {
    expect(shallow(<Dashboard />)).toMatchSnapshot();
    });

    test("it should render blank table component", () => {
      expect(shallow(<Table columns={DashboardTableColumns} tableData={[]} onEdit={jest.fn()} children={<></>} />)).toMatchSnapshot();
    })

    test("it should render table component with Data", () => {
      expect(shallow(<Table columns={DashboardTableColumns} tableData={mockPagedData} onEdit={jest.fn()} children={<></>} />)).toMatchSnapshot();
    })

    test("it should render pagination component", () => {
      expect(shallow(<Pagination currentPage={1} handlePageChange={jest.fn()}/>)).toMatchSnapshot();
    })

    test("it should render form component", () => {
      expect(shallow(<Form data={mockPagedData} handleEdit={jest.fn()} onAdd={jest.fn()} isEdit={true} />)).toMatchSnapshot();
    })
});
