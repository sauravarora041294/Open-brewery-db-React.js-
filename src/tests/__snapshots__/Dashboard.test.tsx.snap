// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard page it should render blank table component 1`] = `
<table
  className="Table"
>
  <thead
    className="TableHeader"
  >
    <tr>
      <th>
        Brewery Name
      </th>
      <th>
        Brewery Type
      </th>
      <th>
        City
      </th>
      <th>
        Country
      </th>
      <th>
        Rating
      </th>
      <th />
    </tr>
  </thead>
  <tbody
    className="TableBody"
  />
</table>
`;

exports[`Dashboard page it should render dashboard component 1`] = `
<div
  className="MainSection"
>
  <h2>
    Open Brewery DataBase
  </h2>
  <div
    className="TableSection"
  >
    <div
      className="TableWrapper"
    >
      <Table
        columns={
          Array [
            "Brewery Name",
            "Brewery Type",
            "City",
            "Country",
            "Rating",
            "",
          ]
        }
        onEdit={[Function]}
        tableData={Array []}
      />
      <Pagination
        currentPage={1}
        handlePageChange={[Function]}
      />
    </div>
    <div
      className="FormWrapper"
    >
      <Form
        data={null}
        handleEdit={[Function]}
        isEdit={false}
        onAdd={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`Dashboard page it should render form component 1`] = `
<div>
  <h2>
    Add/Edit Brewery Details
  </h2>
  <form
    className="FormSection"
  >
    <div
      className="FormRow"
    >
      <label>
        Brewery Name
      </label>
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="FormRow"
    >
      <label>
        Brewery Type
      </label>
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="FormRow"
    >
      <label>
        City
      </label>
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="FormRow"
    >
      <label>
        Country
      </label>
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="FormRow"
    >
      <label>
        Rating
      </label>
      <input
        onChange={[Function]}
        type="number"
        value=""
      />
    </div>
  </form>
  <button
    className="FormButton"
    disabled={true}
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;

exports[`Dashboard page it should render pagination component 1`] = `
<div
  className="PaginationSection"
>
  <button
    className="PaginateButton"
    disabled={true}
    onClick={[Function]}
  >
    Previous
  </button>
  <span>
    1
  </span>
  <button
    className="PaginateButton"
    onClick={[Function]}
  >
    Next
  </button>
</div>
`;

exports[`Dashboard page it should render table component with Data 1`] = `
<table
  className="Table"
>
  <thead
    className="TableHeader"
  >
    <tr>
      <th>
        Brewery Name
      </th>
      <th>
        Brewery Type
      </th>
      <th>
        City
      </th>
      <th>
        Country
      </th>
      <th>
        Rating
      </th>
      <th />
    </tr>
  </thead>
  <tbody
    className="TableBody"
  />
</table>
`;
